### PR TITLE
Pilot RPC

### PR DESCRIPTION
**Pull request by andre-merzky**
This PR implements a pilot RPC call, like this:

```py
        print(pilot.rpc(rpc='hello', args=['world']))
```

which prints:

```sh
hello world
```

The main purpose is to support a use case in the ScaleMS project, where a request to prepare a specific task environment is to be sent to a pilot, and the application needs confirmation that the request succeeded (the existing `prepare_env` call on the pilot does not provide that feedback - it would only fail any later submitted tasks which happen to request an invalid env).

The PR also removes the special pilot staging directory `staging_area`, and instead stages to the pilot sandbox directly.  That change crept into this branch prematurely, and will arrive in devel via #2201 -- sorry for that.

@mtitov , @iparask : I would appreciate help with writing tests for this.  The individual methods are fairly simple - but the whole systems is a chain of DB communications, ZMQ messages and callbacks, so testing is more about integration of those parts than the individual calls.  How should we handle this?

### CODE REVIEW

**Comment by andre-merzky** [comment]
PS.: not sure about the remaining linter comments - pylint seems to ignore the respective `noqa` directives?

File: src/radical/pilot/task_overlay/worker.py
Comment (iparask) [inline src/radical/pilot/task_overlay/worker.py:34]: Line 2 imports os. There is no need to import it here.

File: src/radical/pilot/task_overlay/worker.py
Comment (iparask) [inline src/radical/pilot/task_overlay/worker.py:242]: If you add `# pylint:disable= eval-used` and `# pylint:enable= eval-used` after this line the linter will silence.

File: src/radical/pilot/db/database.py
Comment (iparask) [inline src/radical/pilot/db/database.py:264]: Can we add a test for this method?

**Review by iparask — CHANGES_REQUESTED** [review]

File: src/radical/pilot/task_overlay/worker.py
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:34]: For some reason,  the `os` module was not in the namespace after fork/exec.  Don't ask, but thus the import and the comment...

File: src/radical/pilot/task_overlay/worker.py
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:242]: Any idea why `noqa` does not work?   I am not sure if we should add two lines of linter control for one line of correct code.  Maybe disable that one for the file then?

File: src/radical/pilot/db/database.py
Comment (andre-merzky) [reply src/radical/pilot/db/database.py:264]: Should that be an integration test then?    The logic and mechanics of the method is all around pushing something to MongoDB and waiting for a reply to pop up.  If MongoDB is mocked away, there is not much to test.  Should I convert some RPC example to a test, would that work in your opinion?

File: src/radical/pilot/agent/agent_0.py
Comment (iparask) [inline src/radical/pilot/agent/agent_0.py:611]: Would you mind if I try to create a unit test for this method? I think I have an idea of how to do it.

Comment (andre-merzky) [reply src/radical/pilot/agent/agent_0.py:611]: > Would you mind if I try to create a unit test for this method?

Would I mind?   WOULD *I* MIND?  I thought you'd knew me better than that... :-P

File: src/radical/pilot/db/database.py
Comment (andre-merzky) [reply src/radical/pilot/db/database.py:264]: Makes sense IMHO.

Comment (andre-merzky) [reply src/radical/pilot/db/database.py:264]: That test has been added.

File: src/radical/pilot/task_overlay/worker.py
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py]: should be fixed...

File: src/radical/pilot/agent/agent_0.py
Comment (iparask) [reply src/radical/pilot/agent/agent_0.py:611]: Done

File: tests/test_agent_0/test_unit/test_agent_0.py
Comment (andre-merzky) [inline tests/test_agent_0/test_unit/test_agent_0.py:64]: arg should be a list (`['World']`)

File: tests/test_agent_0/test_unit/test_agent_0.py
Comment (andre-merzky) [inline tests/test_agent_0/test_unit/test_agent_0.py:67]: The print can probably go.

File: tests/test_agent_0/test_unit/test_agent_0.py
Comment (andre-merzky) [inline tests/test_agent_0/test_unit/test_agent_0.py:72]: This will become `hello World` with the above change to list.

**Review by andre-merzky — COMMENTED** [review]
Thanks for adding that test! :-)

**Comment by andre-merzky** [comment]
@iparask : I did some small changes on the test you added.  Please give this PR another review - anything amiss before we can merge?

File: src/radical/pilot/staging_directives.py
Comment (iparask) [inline src/radical/pilot/staging_directives.py:205]: Does this mean that the pilot sandbox gets renamed? Is it not `session_id/pilot.0000/sandbox` anymore? This is something that goes all over the PR.

**Review by iparask — APPROVED** [review]
I'm fine with this PR. It looks okay to me. The linters fail because of archived executing components. @andre-merzky do we want them linted?

If yes, fixing them should be fast. If not, does it make sense to exclude them?

The following line:
```
$(echo $DIFF | grep -o -e '\b[^ ]*.py\b' | sed "s|^|$PWD/|")
```
filters the python files. Should we update it?

**Review by mtitov — APPROVED** [review]
LGTM!

**Comment by andre-merzky** [comment]
We should exclude `archive`'d files from linting (we did the same in travis).

File: src/radical/pilot/staging_directives.py
Comment (andre-merzky) [reply src/radical/pilot/staging_directives.py:205]: Oh no - the pilot sandbox remains as it is.  The special `staging_area/` subdir within the pilot sandbox gets removed though.

Comment (iparask) [reply src/radical/pilot/staging_directives.py:205]: And any data are moved to the same folder with all the log files?

Comment (andre-merzky) [reply src/radical/pilot/staging_directives.py:205]: Yes, indeed.  I realize that this is not optimal, and I suggest to succesively move log, prof and cfg/json files out of the pilot sandbox.  Ideally, the user should be able to inspect the sandbox and mostly recognize workload related data and directories.  But the motivation for this specific change was that it turned out to be difficult to construct relative paths in relation to the `staging_area/`, and the change simplifies sandbox references a lot: `pilot://` refers to the pilot sandbox, `unit://` to the unit sandbox, etc.  This in turn simplifies data management for several projects (scalems, repex, covid).

**Pull request merged**
